modules/tests/model_based.rs

```diff
@@ -0,0 +1,269 @@
+        Height::new(Self::revision(), height)
+    }
+
+    fn mock_header(height: u64) -> MockHeader {
+        MockHeader(Self::height(height))
+    }
+
+    fn header(height: u64) -> AnyHeader {
+        AnyHeader::Mock(Self::mock_header(height))
+    }
+
+    fn client_state(height: u64) -> AnyClientState {
+        AnyClientState::Mock(MockClientState(Self::mock_header(height)))
+    }
+
+    fn consensus_state(height: u64) -> AnyConsensusState {
+        AnyConsensusState::Mock(MockConsensusState(Self::mock_header(height)))
+    }
+
+    fn signer() -> AccountId {
+        AccountId::new([0; 20])
+    }
+
+    /// Check that chain heights match the ones in the model.
+    fn check_chain_heights(&self, chains: HashMap<String, Chain>) -> bool {
+        chains.into_iter().all(|(chain_id, chain)| {
+            let ctx = self.chain_context(chain_id);
+            ctx.query_latest_height() == Self::height(chain.height)
+        })
+    }
+}
+
+impl modelator::TestExecutor<Step> for IBCTestExecutor {
+    fn initial_step(&mut self, step: Step) -> bool {
+        assert_eq!(
+            step.action.action_type,
+            ActionType::None,
+            "unexpected action type"
+        );
+        assert_eq!(
+            step.action_outcome,
+            ActionOutcome::None,
+            "unexpected action outcome"
+        );
+        // initiliaze all chains
+        for (chain_id, chain) in step.chains {
+            self.init_chain_context(chain_id, chain.height);
+        }
+        true
+    }
+
+    fn next_step(&mut self, step: Step) -> bool {
+        let outcome_matches = match step.action.action_type {
+            ActionType::None => panic!("unexpected action type"),
+            ActionType::ICS02CreateClient => {
+                // get action parameters
+                let chain_id = step
+                    .action
+                    .chain_id
+                    .expect("create client action should have a chain identifier");
+                let client_height = step
+                    .action
+                    .client_height
+                    .expect("create client action should have a client height");
+
+                // get chain's context
+                let ctx = self.chain_context_mut(chain_id);
+
+                // create ICS26 message and deliver it
+                let msg = ICS26Envelope::ICS2Msg(ClientMsg::CreateClient(MsgCreateAnyClient {
+                    client_state: Self::client_state(client_height),
+                    consensus_state: Self::consensus_state(client_height),
+                    signer: Self::signer(),
+                }));
+                let result = ctx.deliver(msg);
+
+                // check the expected outcome: client create always succeeds
+                match step.action_outcome {
+                    ActionOutcome::ICS02CreateOK => {
+                        // the implementaion matches the model if no error occurs
+                        result.is_ok()
```

Comment:
should we also check that the new client is stored in the `ctx`, i.e. it can be retrieved?

Reply:
This is a good idea! In the same way we check chain heights after each model action, we can extend it to also check the state of clients. I've opened https://github.com/informalsystems/ibc-rs/issues/629 to track this idea (which is not so easy to implement right now).